graph/src/data/subgraph/schema.rs

```diff
@@ -0,0 +1,343 @@
+//! See `core/src/subgraph/subgraph.graphql` for corresponding graphql schema.
+
+use super::SubgraphId;
+use components::store::{EntityKey, EntityOperation, Store};
+use data::store::Value;
+use failure::Error;
+use std::collections::HashMap;
+use uuid::Uuid;
+
+/// ID of the subgraph of subgraphs.
+pub const SUBGRAPHS_ID: &str = "__subgraphs";
+const EVENT_SOURCE: &str = "SubgraphAdded";
+
+#[derive(Debug)]
+pub struct SubgraphEntity {
+    id: SubgraphId,
+    manifest: SubgraphManifest,
+    created_at: u64,
+}
+
+impl SubgraphEntity {
+    pub fn new(source_manifest: &super::SubgraphManifest, created_at: u64) -> Self {
+        Self {
+            id: source_manifest.id.clone(),
+            manifest: SubgraphManifest::from(source_manifest),
+            created_at,
+        }
+    }
+
+    pub fn write_to_store(self, store: &impl Store) -> Result<(), Error> {
```

Comment:
I deployed the same subgraph a few times (I think) and somehow ended up with duplicate entities in the store:

![image](https://user-images.githubusercontent.com/19324/48550139-2cf79780-e8d2-11e8-9bd4-32548d6fe90e.png)

Reply:
One thing I would do (may not be related) is used deterministic instead of random IDs for all entities. I.e.

* `<subgraph ID>-schema`
* `<subgraph ID>-data-source-0`
* `<subgraph ID>-data-source-0-abi-0`
* `<subgraph ID>-data-source-1`

etc. Unless we're already protecting against writing the same subgraph to the DB twice?

Reply:
Good catch. We have a check for duplicate deployments, but not for writing the subgraph entity twice, which is something we want to be able to do, for example after upgrading `subgraphs.graphql`. Deterministic IDs for all subentities seems like a robust fix, I'll try that.